src/utils/openCloseComponent.ts

```diff
@@ -8,20 +8,67 @@ export interface OpenCloseComponent {
    */
   readonly el: HTMLElement;
 
+  /**
+   * When true, the component opens.
+   */
+  open: boolean;
+
+  /**
+   * Specifies property on which active transition is watched for.
+   */
+  openTransitionProp: string;
+
+  /**
+   * Specifies element that the transition is allowed to emit on.
+   */
+  transitionEl: HTMLDivElement;
+
   /**
    * Defines method for `beforeOpen` event handler.
    */
   onBeforeOpen: () => void;
+
   /**
    * Defines method for `open` event handler:
    */
   onOpen: () => void;
+
   /**
    * Defines method for `beforeClose` event handler:
    */
   onBeforeClose: () => void;
+
   /**
    * Defines method for `close` event handler:
    */
   onClose: () => void;
 }
+
+function transitionStart(event: TransitionEvent): void {
+  if (event.propertyName === this.openTransitionProp && event.target === this.transitionEl) {
+    this.open ? this.onBeforeOpen() : this.onBeforeClose();
+  }
+}
+
+function transitionEnd(event: TransitionEvent): void {
+  if (event.propertyName === this.openTransitionProp && event.target === this.transitionEl) {
+    this.open ? this.onOpen() : this.onClose();
+  }
+}
+
+let boundOnTransitionStart;
```

Comment:
You need to keep a map of listeners. Otherwise, `disconnectOpenCloseComponent` could end up not removing listeners properly. For example:

1. `openCloseComp1` (oCC1) connects
  * `boundOnTransitionStart/End` is updated and references functions created for oCC1
  * listeners for oCC1 added to oCC1
2. `openCloseComp2` (oCC2) connects
  * `boundOnTransitionStart/End` is updated and references functions created for oCC2
  * listeners for oCC2 added to oCC2
3. `openCloseComp1` disconnects
  * `boundOnTransitionStart/End` **still references functions created for oCC2** ⚠️
  * attempts removing listeners for oCC2 from oCC1 – no listeners get removed

Reply:
Yeah, that's a good point. I'll map the listeners.